tests/unit/onyxTest.js

```diff
@@ -500,11 +501,40 @@ describe('Onyx', () => {
                 return waitForPromisesToResolve();
             })
             .then(() => {
-                // THEN the callback should be triggered only once
+                // THEN we expect the callback to be called only once and the initial stored value to be initialCollectionData
                 expect(mockCallback.mock.calls.length).toBe(1);
+                expect(mockCallback.mock.calls[0][0]).toEqual(initialCollectionData);
+            });
+    });
+
+    it('should return all collection keys as a single object when updating a collection key with waitForCollectionCallback = true', () => {
+        const mockCallback = jest.fn();
+        const collectionUpdate = {
+            test_policy_1: {ID: 234, value: 'one'},
+            test_policy_2: {ID: 123, value: 'two'},
+        };
+
+        // GIVEN an Onyx.connect call with waitForCollectionCallback=true
+        connectionID = Onyx.connect({
+            key: ONYX_KEYS.COLLECTION.TEST_POLICY,
+            waitForCollectionCallback: true,
+            callback: mockCallback,
+        });
+        return waitForPromisesToResolve()
+            .then(() => {
+                // WHEN we update the collection, e.g. the API returns a response
+                Onyx.mergeCollection(ONYX_KEYS.COLLECTION.TEST_POLICY, collectionUpdate);
+                return waitForPromisesToResolve();
+            })
+            .then(() => {
+                // THEN we expect the callback to have called twice, once for the initial connect call + once for the collection update
+                expect(mockCallback.mock.calls.length).toBe(2);
+
+                // AND the value for the first call should be null since the collection was not initialized at that point
```

Comment:
Does it seem weird to anyone else that:

1. There would be two calls
2. The first call passes `null`
3. The second call is really what we want with `waitForCollectionCallback=true`

?

Is it possible for us to remove that first call? What is the value in having the call triggered with `null` before the collection has been "initialized"?

Reply:
I think this is expected since we want to initialize that key's value to `null`, since we are connecting to that key, we just don't have a value for it. We leave it up to the subscriber to decide what to do with a null value. 

This happens [here](https://github.com/Expensify/react-native-onyx/blob/2eb3d74e7b581007d7c4dcc332b5bbc027145ce2/lib/Onyx.js#L440-L444) before we run any callback subscriber logic. I think it would make sense to keep the behavior consistent.

Reply:
+1 to what @luacmartins said :), I would also prefer to avoid exceptions and let the subscriber handle that. What if `null` is a valid value and we want to be notified about it?

Reply:
Yeah, those are good points. I think it would be a change in behavior that could make other bugs in the process, so this was more of an exploratory question. Let's leave it as-is!